Applied memory leak fixes

e3m3 opened the pull request:
Closes #13.

e3m3 committed:
Applied memory leak fixes (#13)

- See [discussion](https://github.com/ILLIXR/ILLIXR/pull/112#issuecomment-837308821)

e3m3 committed:
Updated 'libspatialaudio' ('cassert') (#1)

e3m3 commented on src/audio.cpp:
```diff
@@ -57,38 +40,38 @@ void ILLIXR_AUDIO::ABAudio::loadSource(){
+            .fDistance  = 1
+        });
+
+        soundSrcs.emplace_back("samples/radioMusicSample.wav", NORDER, true);
+        soundSrcs.back().setSrcPos({
+            .fAzimuth   = 1.0f, /// Should be same as lectureSample?
+            .fElevation = 0f,   /// Should be same as lectureSample?
+            .fDistance  = 5     /// Should be same as lectureSample?
+        });
```
```suggestion
        soundSrcs.emplace_back("samples/radioMusicSample.wav", NORDER, true);
        soundSrcs.back().setSrcPos({
            .fAzimuth   = 1.0f,
            .fElevation = 0f,
            .fDistance  = 5
        });
```

e3m3 committed:
Removed note from 'src/audio.cpp' (#13)

mhuzai commented on include/sound.h:
```diff
@@ -1,42 +1,39 @@
+#include <spatialaudio/Ambisonics.h>
+
+constexpr std::size_t SAMPLERATE {48000U};
+constexpr std::size_t BLOCK_SIZE {512U};   /// Change to 1024U?
```
512 performs better on Intel's 32 KB data cache.

mhuzai commented on include/sound.h:
```diff
@@ -1,42 +1,39 @@
+constexpr std::size_t SAMPLERATE {48000U};
+constexpr std::size_t BLOCK_SIZE {512U};   /// Change to 1024U?
+constexpr std::size_t NORDER     {3U};
+constexpr std::size_t NUM_SRCS   {16U};    /// Change to 4U?
```
16 is good. It amortizes startup costs during profiling.

mhuzai commented on src/audio.cpp:
```diff
@@ -1,53 +1,36 @@
-        printf("Binauralizer Configuration failed!\n");
-    }
+    /// Binauralizer as ambisonics decoder
+    assert(decoder.Configure(NORDER, true, SAMPLERATE, BLOCK_SIZE, tailLength));
```
I'm not sure if we should remove error checking from `opt` builds. Also, add failure message.

mhuzai commented on src/audio.cpp:
```diff
@@ -1,53 +1,36 @@
-    processType = procTypeIn;
+ILLIXR_AUDIO::ABAudio::ABAudio(std::string outputFilePath, ProcessType procTypeIn)
+    : ProcessType {procTypeIn}
+    , outputFile {
```
Should this be closed in a destructor to make sure all the writes have been flushed to disk?

e3m3 replied:
[Looks like the destructor will take care of the flush](https://stackoverflow.com/questions/4802494/do-i-need-to-close-a-stdfstream).

mhuzai commented on include/sound.h:
```diff
@@ -1,42 +1,39 @@
-
 		Sound(std::string srcFile, unsigned nOrder, bool b3D);
 		// set sound src position
-		void setSrcPos(PolarPoint& pos);
```
Why make a copy?

e3m3 replied:
Maybe there was a overwritten data in a previous implementation? Will change it back.

mhuzai reviewed (commented):
Since I can't comment on submodule diffs:
- I don't understand [this check](https://github.com/ILLIXR/libspatialaudio/compare/380107c4a6b0c26e20281b6f919092a9d8411c56...9bfd57756d8dd3c9194bdf931671b8265dc2b0d2#diff-97faad21c9f99856e1c6c19ca137c4dc11c65bf915fa2026c0f5ede03145c974R62). There is no pointer aliasing going on here, and the compiler/OS isn't going to allocate memory that overwrites a class member. So why is that check here?
- Why are there two of [these](https://github.com/ILLIXR/libspatialaudio/compare/380107c4a6b0c26e20281b6f919092a9d8411c56...9bfd57756d8dd3c9194bdf931671b8265dc2b0d2#diff-97faad21c9f99856e1c6c19ca137c4dc11c65bf915fa2026c0f5ede03145c974R73)?

e3m3 commented:
>    I don't understand this check. There is no pointer aliasing going on here, and the compiler/OS isn't going to allocate memory that overwrites a class member. So why is that check here?

I agree. If a previous implementation used a static value for the buffer size then we would need it, but the `m_pfDelayBuffer` buffer is now heap allocated. Removing.

>    Why are there two of these?

If aliasing were to occur like the check above mentions, the `memset` could write a bogus value to `m_nDelayBufferLength`. Removing.

e3m3 committed:
Addressing reviewer comments (#13)

- Added abort routine 'configAbort' in 'include/audio.h'
- Removed parameter notes from 'include/sound.h'
- Changed 'setSrcPos' to call by reference in 'include/sound.h'
- Added error checks for configurations in 'source/audio.cpp'
- Changed 'setSrcPos' to call by reference in 'source/sound.cpp'
- Implemented 'configAbort' in 'source/audio.cpp'
- Updated 'libspatialaudio'

e3m3 committed:
Added missing 'ILLIXR_INTEGRATION' in 'Makefile' (#13)

e3m3 committed:
Added missing 'ILLIXR_INTEGRATION' in 'Makefile' (#13)

e3m3 committed:
Fixing compilation bugs (#13)

- Fixed 'configAbort' signature in 'include/audio.h'
- Fixed source initialization in 'src/audio.cpp'
- Fixed source loop in 'src/audio.cpp'
- Fixed 'configAbort' implementation in 'src/audio.cpp'

e3m3 committed:
Fixing compilation issues (#13)

- Removed unused 'position' in 'src/audio.cpp'
- Fixed initialization of 'processType' in 'src/audio.cpp'

e3m3 committed:
Fixing compilation issue (#13)

- Added missing 'cstddef' include in 'src/sound.cpp'
- Fixed dereferencing for 'BFormat' in 'src/sound.cpp'
- Fixed 'srcPos' initialization in 'src/sound.cpp'
- Fixed dereferencing for 'srcFile' in 'src/sound.cpp'

e3m3 committed:
Fixing compilation error (#13)

- Fixed call to 'reinterpret_cast' in 'src/sound.cpp'
- Fixed call to 'BFormat.get' in 'src/sound.cpp'

e3m3 committed:
Fixed call to 'reinterpret_cast' in 'src/sound.cpp' (#13)

e3m3 committed:
Fixing compilation issues (#13)

- Moved include of 'optional' : 'src/audio.cpp' -> 'include/audio.h'

e3m3 committed:
Updated libspatialaudio (#13)

e3m3 committed:
Updated libspatialaudio (#13)

e3m3 committed:
Updated libspatialaudio (#13)

e3m3 committed:
Fixing sound source double-free (#13)

- Changed 'BFormat' member to 'std::shated_ptr' in 'include/sound.h'
- Changed 'readInBFormat' return type to 'std::weak_ptr' in 'include/sound.h'
- Updated 'readNEncode' to use 'std::weak_ptr' in 'src/audio.cpp'
- Updated 'BFormat' implementation in 'src/sound.cpp'

e3m3 committed:
Fixed call to 'ILLIXR::abort' in 'src/audio.cpp' (#13)

e3m3 committed:
Fixing error checking (#13)

- Added abort routine 'configAbort' to 'include/sound.h'
- Updated error checking for 'BFormat' in 'src/sound.cpp'
- Updated error checking for 'BEncode' in 'src/sound.cpp'
- Added 'configAbort' implementation to 'src/sound.cpp'

e3m3 committed:
Updated libspatialaudio (#13)

e3m3 commented:
Ready for review
